pkg/protocol/engine/utxoledger/snapshot.go

```diff
@@ -90,134 +79,136 @@ func SpentFromSnapshotReader(reader io.ReadSeeker, apiProvider iotago.APIProvide
 		return nil, err
 	}
 
-	transactionIDSpent := iotago.TransactionID{}
-	if _, err := io.ReadFull(reader, transactionIDSpent[:]); err != nil {
-		return nil, ierrors.Errorf("unable to read LS transaction ID spent: %w", err)
+	transactionIDSpent, err := stream.Read[iotago.TransactionID](reader)
+	if err != nil {
+		return nil, ierrors.Wrap(err, "unable to read LS transaction ID spent")
 	}
 
 	return NewSpent(output, transactionIDSpent, indexSpent), nil
 }
 
 func ReadSlotDiffToSnapshotReader(reader io.ReadSeeker, apiProvider iotago.APIProvider) (*SlotDiff, error) {
+	var err error
 	slotDiff := &SlotDiff{}
 
-	var diffIndex iotago.SlotIndex
-	if err := binary.Read(reader, binary.LittleEndian, &diffIndex); err != nil {
-		return nil, ierrors.Errorf("unable to read slot diff index: %w", err)
+	if slotDiff.Slot, err = stream.Read[iotago.SlotIndex](reader); err != nil {
+		return nil, ierrors.Wrap(err, "unable to read slot diff index")
 	}
-	slotDiff.Slot = diffIndex
 
-	var createdCount uint64
-	if err := binary.Read(reader, binary.LittleEndian, &createdCount); err != nil {
-		return nil, ierrors.Errorf("unable to read slot diff created count: %w", err)
+	createdCount, err := stream.Peek(reader, serializer.SeriLengthPrefixTypeAsUint64)
+	if err != nil {
+		return nil, ierrors.Wrap(err, "unable to peek slot diff created count")
 	}
-
 	slotDiff.Outputs = make(Outputs, createdCount)
 
-	for i := uint64(0); i < createdCount; i++ {
-		var err error
+	if err := stream.ReadCollection(reader, serializer.SeriLengthPrefixTypeAsUint64, func(i int) error {
 		slotDiff.Outputs[i], err = OutputFromSnapshotReader(reader, apiProvider)
 		if err != nil {
-			return nil, ierrors.Errorf("unable to read slot diff output: %w", err)
+			return ierrors.Wrap(err, "unable to read slot diff output")
 		}
-	}
 
-	var consumedCount uint64
-	if err := binary.Read(reader, binary.LittleEndian, &consumedCount); err != nil {
-		return nil, ierrors.Errorf("unable to read slot diff consumed count: %w", err)
+		return nil
+	}); err != nil {
+		return nil, ierrors.Wrap(err, "unable to read slot diff created collection")
 	}
 
+	consumedCount, err := stream.Peek(reader, serializer.SeriLengthPrefixTypeAsUint64)
```

Comment:
we can change this to uint32 to match SlotDiff. Don't think we will have this many outputs consumed in a single slot